File: javapatterns/src/main/java/org/iut/nantes/state/canards/Canard.java
package org.iut.nantes.state.canards;

import org.iut.nantes.state.canards.etats.*;

public class Canard {
    private Etat sousLeau = new SousLeau(this);
    private Etat dansLair = new DansLair(this);
    private Etat surLeau = new SurLeau(this);
    private Etat surTerre = new SurTerre(this);

    private Etat etatCourant = surLeau;

    public void avancer() {
        etatCourant.avancer();
    }

    public void palmeATerre() {
        etatCourant.palmeATerre();
    }

    public void trempette() {
        etatCourant.trempette();
    }

    public void plongeon() {
        etatCourant.plongeon();
    }

    public void decollage() {
        etatCourant.decollage();
    }

    public void changeEtatSurLeau() {
        etatCourant = surLeau;
    }

    public void changeEtatSousLeau() {
        etatCourant = sousLeau;
    }

    public void changeEtatSurTerre() {
        etatCourant = surTerre;
    }

    public void changeEtatDansLair() {
        etatCourant = dansLair;
    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/state/canards/Canard.kt
package org.iut.nantes.state.canards

import org.iut.nantes.state.canards.etats.*

class Canard {

    private val etatSurEau: Etat = SurLeau(this)
    private val etatDansAir: Etat = DansLair(this)
    private val etatSurTerre: Etat = SurTerre(this)
    private val etatDansEau: Etat = SousLeau(this)

    var etat: Etat = etatSurEau

    fun avancer() {
        etat.avancer()
    }

    fun plongeon() {
        etat.plongeon()
    }

    fun decollage() {
        etat.decollage()
    }

    fun palmeATerre() {
        etat.palmeATerre()
    }

    fun trempette() {
        etat.trempette()
    }

    fun changeEtatSurEau() {
        etat = etatSurEau
    }

    fun changeEtatDansAir() {
        etat = etatDansAir
    }

    fun changeEtatSurTerre() {
        etat = etatSurTerre
    }

    fun changeEtatDansEau() {
        etat = etatDansEau
    }
}